### ISSUE TITLE
Quick Question about Aggregation Network

### ISSUE THREAD

**Issue by jmahajan117** [google-research/readout_guidance#14]
Hey,

Awesome work here. I just had a question. Why do you update the aggregation network once, instead of every timestep?

Wouldn't be more efficient to update the layers and a set amount of weights for each timestep rather than doing a one large update for the all timesteps?

Thanks!

**Comment by g-luo** [comment]
Hi, could you provide a reference to a line of code or file with this behavior? Specifically, I'm not sure if your question is referring to [training](https://github.com/google-research/readout_guidance/blob/39f0197ddedf9027231ce292360f61d0d21f6cca/readout_training/train_spatial.py#L104) or [guidance](https://github.com/google-research/readout_guidance/blob/39f0197ddedf9027231ce292360f61d0d21f6cca/readout_guidance/rg_operators.py#L189).

**Comment by jmahajan117** [comment]
Hey Grace,

I think I found my answer on that, but I have a couple more. 

How did you choose the architecture of the aggregation network? Did you take some inspiration from some model?

Also, how are you extracting the features from the unet? From what I'm understanding, you encode an input image into the vae and at a random timestep in the diffusion process, you extract the in between outputs of the up blocks and finally you upsample them to a common dimension? Let me know if that's correct.

Thanks,
Jay

**Comment by g-luo** [comment]
Hi, we build on top of the aggregation network from [Diffusion Hyperfeatures](https://diffusion-hyperfeatures.github.io).

Yes, that correctly describes the [forward pass of the DiffusionExtractor](https://github.com/google-research/readout_guidance/blob/39f0197ddedf9027231ce292360f61d0d21f6cca/dhf/diffusion_extractor.py#L137). We extract and cache the features [here](https://github.com/google-research/readout_guidance/blob/39f0197ddedf9027231ce292360f61d0d21f6cca/dhf/stable_diffusion/resnet.py#L82), where we set `save_mode=hidden` by default.

**Comment by jmahajan117** [comment]
Hi, another question, for guidance, if I'm understanding correctly, you optimize on the latents and use the unet & readout network for backpropagation correct? You don't update the parameters of the readout network? And you use the entire unet?

And your loss is between the output of the readout network and the image you want to guide towards?

**Comment by g-luo** [comment]
Hi @jmahajan117,

- Yes, we optimize the latents, which is backpropagated through the readout head and unet. We do not update any model parameters.
- To compute a readout, we aggregate over the hidden activation from each resnet in the decoder half of the unet (see [here](https://github.com/google-research/readout_guidance/blob/39f0197ddedf9027231ce292360f61d0d21f6cca/dhf/stable_diffusion/resnet.py#L100)). Therefore we are indeed using the entire unet for backpropagation. You could theoretically use a smaller subset of the unet if you select a single early unet feature to compute the readout, e.g., layer 4 for the spatial tasks.
- For spatial tasks (pose / depth / edge), yes, [the loss](https://github.com/google-research/readout_guidance/blob/39f0197ddedf9027231ce292360f61d0d21f6cca/readout_guidance/rg_operators.py#L157) is the MSE between the readout and the spatial control you want to guide towards.
